File: panel/frontend/src/features/security/types/security.ts
// Security types — payload of the single /api/security round-trip.
//
// Mirrors internal/models/security.go (Go-side shape). The Security admin
// page renders every field here; the React side never constructs this
// shape from scratch — it only consumes what the handler returns.

export interface SecuritySnapshot {
  generated_at: string;
  // WindowSeconds is the aggregate window the cumulatives cover (the
  // backend uses 5 minutes). Surfaced as a "over the last Ns" suffix so
  // operators reading the numbers understand their scope.
  window_seconds: number;

  // Headline counters — cumulative over the window.
  total_requests: number;
  requests_per_second: number;
  requests_per_minute: number;
  active_connections: number;
  unique_visitors: number;
  blocked_requests: number;
  challenge_count: number;
  errors_4xx: number;
  errors_5xx: number;
  api_requests: number;
  login_attempts: number;
  failed_login_attempts: number;
  bandwidth_bytes: number;
  peak_rps: number;
  average_response_ms: number;
  defense_cpu_pct: number;
  defense_memory_pct: number;
  under_attack: boolean;

  // DDoS runtime state (not editable, surfaced for monitoring).
  ddos_active: boolean;
  ddos_stop_count: number;
  ddos_cooldown_until: string;

  // DDoS port-switch runtime state: which port the panel is actually
  // serving on right now and whether a reaction moved it off the launch
  // port. ddos_port_error carries the last bind failure (e.g. alt port
  // already in use) so the UI can explain why a switch didn't happen.
  ddos_active_port: number;
  ddos_port_switched: boolean;
  ddos_port_error: string;

  // TCP-level DDoS listener counters (process-cumulative): how many
  // connections were accepted vs hard-refused at the socket layer while
  // auto-stop was active.
  ddos_tcp_accepted: number;
  ddos_tcp_dropped: number;

  // Config is the editable runtime config the Security page reads to
  // decide what to display in the rate-limit form. Lives inside the
  // snapshot so the page can paint both the live numbers and the editable
  // controls in a single round-trip; the PUT endpoint echoes the same
  // shape back so the SPA can refresh its local state without re-fetching.
  config: SecurityConfig;

  // Top-N lists — already capped server-side.
  top_ips: SecurityTopEntry[];
  requests_per_ip: SecurityTopEntry[];
  countries: SecurityTopEntry[];
  user_agents: SecurityTopEntry[];
}

export interface SecurityConfig {
  // RequestsPerMinuteLimit caps how many requests a single client IP may
  // make inside the rolling WindowSecondsLimit window. Expressed as
  // requests-per-minute to match the UI label. 0 disables per-IP throttling.
  requests_per_minute_limit: number;
  // WindowSecondsLimit is the size of the rolling window the per-IP cap
  // is evaluated against. Defaults to 60.
  window_seconds_limit: number;
  // GlobalRPMLimit caps total requests-per-minute across the whole panel.
  // When crossed AND UnderAttack is on, every additional request returns
  // 429 + is stamped blocked = 1 in the telemetry table. 0 disables.
  global_rpm_limit: number;
  // BlockUnknownUA causes the middleware to short-circuit any request
  // whose User-Agent header is empty (curl/wget probes, headless scrapers).
  // Off by default to keep scripted tooling working out of the box.
  block_unknown_ua: boolean;

  // DDOSAutoStopEnabled enables automatic DDoS protection: when a single
  // IP exceeds the per-minute limit, or the global burst detector (see
  // ddos_global_trigger_hits) trips, the configured reaction fires
  // (ddos_mode) for ddos_stop_minutes.
  ddos_auto_stop_enabled: boolean;
  // DDOSStopMinutes is how many minutes the panel stays stopped after a
  // DDoS auto-stop triggers. Default 5.
  ddos_stop_minutes: number;
  // DDOSMaxStopCount is the maximum number of times the panel will
  // automatically stop. 0 = unlimited. After reaching this count,
  // auto-stop is disabled until manually reset.
  ddos_max_stop_count: number;
  // DDOSMode selects the reaction when a DDoS is detected:
  //   - "stop": refuse new requests for the cooldown window.
  //   - "port_switch": move the panel onto ddos_alt_port so attackers
  //     keep hitting a dead port while the panel stays reachable.
  // Mirrors models.DDOSMode* constants on the Go side.
  ddos_mode: 'stop' | 'port_switch';
  // DDOSAltPort is the port the panel moves to in "port_switch" mode.
  // Must be 1-65535 and differ from the current panel port.
  ddos_alt_port: number;
  // Global burst detector: trips auto protection when more than this
  // many requests arrive across ALL IPs inside the window below (catches
  // distributed floods no single IP would trip). 0 = disabled.
  ddos_global_trigger_hits: number;
  // Sliding window (seconds) for the global burst detector. Clamped to
  // 5-60 by the backend.
  ddos_global_trigger_window: number;

  // ── Firewall tab: IP allow/deny + WAF / Request Filtering ──
  // Client IPs or CIDRs that bypass the per-IP rate limit.
  ip_allowlist: string[];
  // Client IPs or CIDRs that are hard-rejected with 403.
  ip_denylist: string[];
  // Per-request body size cap in megabytes (backend clamps >= 1).
  max_body_size_mb: number;
  // CSV HTTP-method allowlist ("" = every method allowed).
  allowed_http_methods: string;
  // Reject known scanner/probe paths (/.env, /wp-admin, ...) with 403.
  block_suspicious_paths: boolean;

  // ── Sessions tab: session policy ──
  // Absolute session lifetime in minutes (drives cookie TTL + bearer
  // max-age). Default 480 (= previous hardcoded 8h).
  session_lifetime_minutes: number;
  // Idle timeout in minutes for tracked sessions (default 1440 = 24h).
  session_idle_timeout_minutes: number;
  // Maximum concurrent active sessions per user (0 = unlimited).
  session_max_per_user: number;
}

// One tracked session on the Sessions tab's devices table. `id` is a
// non-reversible SHA-256 prefix of the token used as the revoke handle —
// full tokens never leave the server.
export interface SecuritySessionEntry {
  id: string;
  user_id: number;
  username: string;
  ip_address: string;
  user_agent: string;
  issued_at: string;
  last_used: string;
  current: boolean;
}

export interface SecuritySessionsResponse {
  sessions: SecuritySessionEntry[];
  total: number;
}

export interface CookieSecurityInfo {
  name: string;
  host_prefix: boolean;
  http_only: boolean;
  same_site: string;
  secure: boolean;
  path: string;
  lifetime_min: number;
  idle_timeout_min: number;
}

// Read-only snapshot of panel-wide network protections rendered by the
// Firewall tab's status cards. The backend reports what is actually wired,
// including honest "not enforced" answers.
export interface SecurityStatusResponse {
  cors: {
    credentials: boolean;
    development_mode: boolean;
    origin_validation: boolean;
    allowed_origins: string[];
  };
  csrf: {
    token_middleware_enforced: boolean;
    session_cookie_same_site: string;
    origin_validation: boolean;
    note: string;
  };
  security_headers: {
    enforced: boolean;
    middleware_wired: boolean;
    applied_headers: string[];
    note: string;
  };
  request_limits: {
    max_body_mb: number;
  };
  cookie: CookieSecurityInfo;
}

// Login-protection status (Authentication tab): the effective lockout
// policy plus currently locked accounts.
export interface LockoutStatus {
  max_attempts: number;
  window_minutes: number;
  lockout_minutes: number;
  locked: Array<{ username: string; locked_at: string }>;
}

export interface RecoveryCodesStatus {
  users_with_codes: number;
  unused_codes: number;
}

export interface SecurityTopEntry {
  label: string;
  count: number;
}

File: panel/backend/internal/models/security.go
package models

import "time"

// SecuritySnapshot is the single response shape returned by the Security
// admin page (admin/security). It folds the per-request security_requests
// feed into the headline counters + top-N lists an operator needs to
// monitor the panel under attack, in one round-trip so the page can paint
// in a single fetch and refresh on an interval without re-querying.
//
// Every aggregate field is computed up-front by SecurityRepository.Snapshot
// against a configurable window (last 60s for RPS, full-window for cumulatives).
// The "Peak RPS" figure is the highest per-second bucket seen in the window.
type SecuritySnapshot struct {
	GeneratedAt time.Time `json:"generated_at"`
	// WindowSeconds is the aggregate window the cumulatives cover. The
	// dashboard surfaces it as the "over the last Ns" suffix on each tile
	// so an operator reading the numbers understands their scope.
	WindowSeconds int64 `json:"window_seconds"`

	// --- Headline counters (cumulative over the window) ---
	TotalRequests       int64 `json:"total_requests"`
	RequestsPerSecond   int64 `json:"requests_per_second"`
	RequestsPerMinute   int64 `json:"requests_per_minute"`
	ActiveConnections   int64 `json:"active_connections"`
	UniqueVisitors      int64 `json:"unique_visitors"`
	BlockedRequests     int64 `json:"blocked_requests"`
	ChallengeCount      int64 `json:"challenge_count"`
	Errors4xx           int64 `json:"errors_4xx"`
	Errors5xx           int64 `json:"errors_5xx"`
	APIRequests         int64 `json:"api_requests"`
	LoginAttempts       int64 `json:"login_attempts"`
	FailedLoginAttempts int64 `json:"failed_login_attempts"`
	BandwidthBytes      int64 `json:"bandwidth_bytes"`
	PeakRPS             int64 `json:"peak_rps"`
	AverageResponseMs   int64 `json:"average_response_ms"`
	DefenseCPU          int64 `json:"defense_cpu_pct"`
	DefenseMemory       int64 `json:"defense_memory_pct"`
	UnderAttack         bool  `json:"under_attack"`

	// DDoS runtime state (not editable, surfaced for monitoring).
	DDOSActive        bool   `json:"ddos_active"`
	DDOSStopCount     int64  `json:"ddos_stop_count"`
	DDOSCooldownUntil string `json:"ddos_cooldown_until"`

	// DDoS port-switch runtime state. DDOSActivePort is the port the
	// panel is actually bound to right now (equals the launch port when
	// no switch happened); DDOSPortSwitched is true while the panel is
	// serving on DDOSAltPort because of an active attack; DDOSPortError
	// carries the last bind failure so the admin can see WHY a switch
	// didn't take effect (e.g. alt port already in use).
	DDOSActivePort   int64  `json:"ddos_active_port"`
	DDOSPortSwitched bool   `json:"ddos_port_switched"`
	DDOSPortError    string `json:"ddos_port_error"`

	// DDOSTCPDropped counts how many TCP connections the listener has
	// refused at the socket layer since launch because the DDoS
	// auto-stop was active. The counter is read from the wrapper in
	// internal/api/ddos_listener.go. It is process-cumulative, not
	// window-bound, so the admin page shows the total impact of the
	// current auto-stop (and prior stops) at a glance.
	DDOSTCPDropped  uint64 `json:"ddos_tcp_dropped"`
	DDOSTCPAccepted uint64 `json:"ddos_tcp_accepted"`

	// Config is the editable runtime config the Security page reads to
	// decide what to display in the rate-limit form. Lives inside the
	// snapshot so the page can paint both the live numbers and the editable
	// controls in a single round-trip; the PUT endpoint echoes the same
	// shape back so the SPA can refresh its local state without re-fetching.
	Config SecurityConfig `json:"config"`

	// --- Top-N lists (already capped server-side) ---
	TopIPs        []SecurityTopEntry `json:"top_ips"`
	RequestsPerIP []SecurityTopEntry `json:"requests_per_ip"`
	Countries     []SecurityTopEntry `json:"countries"`
	UserAgents    []SecurityTopEntry `json:"user_agents"`
}

// SecurityConfig is the operator-editable subset of the security policy. It
// is persisted in the settings KV (migration 028) and read by both the
// SecurityMiddleware (to decide which requests to block) and the Security
// admin page (to render the edit form). Every field has a safe default
// loaded when the row is missing so a brand-new install behaves sensibly
// without the admin ever opening the page.
type SecurityConfig struct {
	// RequestsPerMinuteLimit caps how many requests a single client IP may
	// make inside the rolling `WindowSecondsLimit` window. Expressed as
	// requests-per-minute to match the UI label. 0 disables per-IP throttling.
	RequestsPerMinuteLimit int64 `json:"requests_per_minute_limit"`
	// WindowSecondsLimit is the size of the rolling window the per-IP cap
	// is evaluated against. Defaults to 60.
	WindowSecondsLimit int64 `json:"window_seconds_limit"`
	// GlobalRPMLimit caps total requests-per-minute across the whole panel.
	// When crossed AND UnderAttack is on, every additional request returns
	// 429 + is stamped blocked = 1 in the telemetry table. 0 disables.
	GlobalRPMLimit int64 `json:"global_rpm_limit"`
	// BlockUnknownUA causes the middleware to short-circuit any request
	// whose User-Agent header is empty (curl/wget probes, headless scrapers).
	// Off by default to keep scripted tooling working out of the box.
	BlockUnknownUA bool `json:"block_unknown_ua"`

	// DDOSAutoStopEnabled enables automatic panel stop when DDoS is detected.
	// When on, if a single IP exceeds the per-minute limit while UnderAttack
	// is active, the panel stops accepting new requests for the cooldown period.
	DDOSAutoStopEnabled bool `json:"ddos_auto_stop_enabled"`
	// DDOSStopMinutes is how many minutes the panel stays stopped after a
	// DDoS auto-stop triggers. Default 5.
	DDOSStopMinutes int64 `json:"ddos_stop_minutes"`
	// DDOSMaxStopCount is the maximum number of times the panel will
	// automatically stop. 0 = unlimited. After reaching this count,
	// auto-stop is disabled until manually reset.
	DDOSMaxStopCount int64 `json:"ddos_max_stop_count"`
	// DDOSMode selects what happens when a DDoS attack is detected:
	//   - "stop": refuse new requests for the cooldown window (default).
	//   - "port_switch": close the current panel port and re-bind the
	//     panel on DDOSAltPort, so attackers keep hitting a dead port
	//     while the panel stays reachable on the alternate one.
	DDOSMode string `json:"ddos_mode"`
	// DDOSAltPort is the port the panel moves to when DDOSMode is
	// "port_switch" and an attack is detected. Must be 1-65535 and must
	// differ from the port the panel is currently bound to.
	DDOSAltPort int64 `json:"ddos_alt_port"`
	// DDOSGlobalTriggerHits triggers auto protection when more than this
	// many requests arrive across ALL client IPs inside
	// DDOSGlobalTriggerWindow seconds (a distributed flood that no single
	// IP trips the per-IP cap). 0 disables the global burst detector and
	// leaves detection to the per-IP rate limit path.
	DDOSGlobalTriggerHits int64 `json:"ddos_global_trigger_hits"`
	// DDOSGlobalTriggerWindow is the sliding window in seconds the global
	// burst detector evaluates. Clamped to 5-60 by the PUT handler.
	DDOSGlobalTriggerWindow int64 `json:"ddos_global_trigger_window"`

	// --- Firewall tab: IP allow/deny + WAF request filtering ---
	// IPAllowlist lists client IPs (exact address or CIDR, e.g.
	// "10.0.0.5" or "192.168.1.0/24") that bypass the per-IP rate
	// limit. Empty list = no allowlist; every IP is throttled normally.
	IPAllowlist []string `json:"ip_allowlist"`
	// IPDenylist lists client IPs / CIDRs that are hard-rejected with
	// 403 before any other check runs.
	IPDenylist []string `json:"ip_denylist"`
	// MaxBodySizeMB caps how large a single request body may be.
	// Replaces the previous hardcoded 10 MB middleware constant;
	// clamped to >= 1 by the PUT handler.
	MaxBodySizeMB int64 `json:"max_body_size_mb"`
	// AllowedHTTPMethods is a comma-separated allowlist of HTTP methods
	// (e.g. "GET,POST,PUT,DELETE"). Empty = every method is allowed.
	// OPTIONS preflights are always permitted.
	AllowedHTTPMethods string `json:"allowed_http_methods"`
	// BlockSuspiciousPaths rejects known scanner/probe paths (/.env,
	// /wp-admin, ...) with 403 when enabled.
	BlockSuspiciousPaths bool `json:"block_suspicious_paths"`

	// --- Sessions tab: session policy ---
	// SessionLifetimeMinutes bounds the absolute life of a session
	// (cookie TTL, bearer max-age and rotation window derive from it).
	// 480 mirrors the previous hardcoded 8h auth.SessionTTL.
	SessionLifetimeMinutes int64 `json:"session_lifetime_minutes"`
	// SessionIdleTimeoutMinutes invalidates a tracked session that has
	// not been used for this long. 1440 mirrors the previous hardcoded
	// 24h cleanup window. Applies only to sessions registered in the
	// in-memory SessionManager (cookie logins + switch-login tokens).
	SessionIdleTimeoutMinutes int64 `json:"session_idle_timeout_minutes"`
	// SessionMaxPerUser caps concurrent active sessions per user;
	// the oldest session is evicted first. 0 = unlimited (the default,
	// matching pre-config behaviour).
	SessionMaxPerUser int64 `json:"session_max_per_user"`
}

// DDoS protection modes accepted for SecurityConfig.DDOSMode.
const (
	// DDOSModeStop refuses new requests while an attack is active.
	DDOSModeStop = "stop"
	// DDOSModePortSwitch moves the panel from its current port to
	// SecurityConfig.DDOSAltPort while an attack is active.
	DDOSModePortSwitch = "port_switch"
)

// SecurityTopEntry is one row of any of the Top-N lists (top IPs,
// requests-per-IP, countries, user agents). The `Label` is the dimension
// value (the IP address, the country code, the user-agent string) and
// `Count` is how many requests matched it in the snapshot window.
type SecurityTopEntry struct {
	Label string `json:"label"`
	Count int64  `json:"count"`
}
